File: interpreter.py
lexicon = {'+','-','>','<','.',',','[',']'}

error_occurred = False

def error(message):
    print("\033[91m {}\033[00m" .format(message))
    global error_occurred
    error_occurred = True

class brainfuck:
    def __init__ (self,file):
        self.bytecells = [0] * 30_000
        self.ptr = 0
        self.pos = 0
        self.stack = []
        with open(file, 'r') as file:
            self.content = file.read()
    
    def output(self): print(chr(self.bytecells[self.ptr]), end = '')
    
    def input(self): self.bytecells[self.ptr] = ord(input('\n'))

    def increment(self): self.bytecells[self.ptr] = (self.bytecells[self.ptr] + 1) % 256

    def decrement(self): self.bytecells[self.ptr] = (self.bytecells[self.ptr] - 1) % 256

    def up(self): self.ptr = (self.ptr + 1) % 30_000
    
    def down(self): self.ptr = (self.ptr - 1) % 30_000

    def parser(self, i):
        if i == '+': self.increment()
        if i == '-': self.decrement()
        if i == '>': self.up()
        if i == '<': self.down()
        if i == '.': self.output()
        if i == ',': self.input()

    def execute(self):
        global error_occurred
        while self.pos < len(self.content) and not error_occurred:
            command = self.content[self.pos]
            if command == '[':
                self.stack.append(self.pos)
            elif command == ']':
                if len(self.stack) == 0:
                    error("Closing ']' encountered without previous '['")
                elif self.bytecells[self.ptr] != 0:
                    self.pos = self.stack[-1]
                else:
                    self.stack.pop()
            else:
                self.parser(command)
            self.pos += 1

bf = brainfuck('main.bf')
bf.execute()

File: interpreter.cpp
#include <bits/stdc++.h>
using namespace std;

char lexicon[] = {'+', '-', '>', '<', '.', ',', '[', ']'};

bool error_occurred = false;

void error(const string& message) {
    cerr << "\033[91m" << message << "\033[00m" << endl;
    error_occurred = true;
}

class Brainfuck {
private:
    vector<char> bytecells;
    int ptr;
    int pos;
    stack<int> stack;
    string content;

public:
    Brainfuck(const string& filename)
        : bytecells(30000, 0), ptr(0), pos(0) {
        ifstream file(filename);
        if (!file.is_open()) {
            error("Unable to open file");
            exit(1);
        }
        getline(file, content, '\0');
        file.close();
    }

    void output() { cout << bytecells[ptr]; }

    void input() {
        char c;
        cin >> c;
        bytecells[ptr] = c;
    }

    void increment() { bytecells[ptr] = (bytecells[ptr] + 1) % 256; }

    void decrement() { bytecells[ptr] = (bytecells[ptr] - 1) % 256; }

    void up() { ptr = (ptr + 1) % 30000; }

    void down() { ptr = (ptr - 1) % 30000; }

    void parser(char command) {
        if (command == '+') increment();
        else if (command == '-') decrement();
        else if (command == '>') up();
        else if (command == '<') down();
        else if (command == '.') output();
        else if (command == ',') input();
    }

    void execute() {
        while (pos < content.size() && !error_occurred) {
            char command = content[pos];
            if (command == '[') {
                stack.push(pos);
            } else if (command == ']') {
                if (stack.empty()) {
                    error("Closing ']' encountered without previous '['");
                } else if (bytecells[ptr] != 0) {
                    pos = stack.top();
                } else {
                    stack.pop();
                }
            } else {
                parser(command);
            }
            pos++;
        }
    }
};

int main() {
    Brainfuck bf("main.bf");
    bf.execute();
    return 0;
}
